### PR TITLE
ENHANCEMENT rewrite image tags to image shortcodes

### PR DESCRIPTION
**Pull request by muskie9**
- [x] Update consolidate yield methods to one single dimension and one multi-dimension array methods

This task will query `File` based on the on the final part of the source string `.../../../myImage.jpg`. It replaces the source with the file URL based on the returned record. This doesn't currently handle resampled images, but rather uses the primary url for the file.

ping #167

### CODE REVIEW

**Comment by muskie9** [comment]
There is also a bug with the task where some instances like `title=""`  update to `title=""""`.

**Comment by muskie9** [comment]
I ran some migration tests and if I replace `""` with `\"\"` in the `getNewFieldValue()` method the result when viewing the source in the wysiwyg is `"\"\""` so it seems to wrap it regardless. the `echo` for one such image in the test ended up in this format before the `UPDATE` call:

`[image width="350" height="278" title=\"\" class="right" alt="my image" src="/assets/path/to/my/904fcc50b1/image.jpg"]`

Without the update, as noted above, the result would be `[image width="350" height="278" title="" class="right" alt="my image" src="/assets/path/to/my/904fcc50b1/image.jpg"]` before the `UPDATE`

It seems to wrap empty quotes with quotes during the query.

File: src/ContentFileMigrationTask.php [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    {
+        foreach ($this->getMapping() as $class => $tables) {
+            foreach ($tables as $table => $fields) {
+                $selectString = 'ID';
+
+                foreach ($fields as $field) {
+                    $selectString = "$selectString, \"{$field}\"";
+                }
+
+                $records = DB::query("SELECT {$selectString} FROM \"{$table}\"");
```
Comment (muskie9) [inline src/ContentFileMigrationTask.php]: This query should likely be run on the base, versions and live table (if available) for each class that has an eligible field.

**Comment by muskie9** [comment]
- [ ] Fix instances where Live table gets updated but image isn't rendered when visiting the page

**Comment by muskie9** [comment]
Is there any concern that an image reference might have an `id` for css/js purposes as it relates to the new short code? In a small test run I added `id="my-unique-image-id"` to the legacy image tag and it doesn't seem to like it very much. I'm not sure this would be encountered a whole lot.

I tested in a SS3 instance and the `id="foo-bar-baz"` was recorded but in SS4 it looks like it is stripped out before the save. One option would be to move the `id` if it exists to `class` and note it in upgrade docs, but I'll defer as I don't know if that's something that should be handled at a different level in case a user tries to add an `id` via the wysiwyg editor in SS4.

I also noticed when viewing the SS4 record's html in the wysiwyg, the file ID shows as `data-id` but is then saved in the db as `id`. Curious to hear what others think about this case.

File: src/ContentFileMigrationTask.php [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+     * @param $match
+     * @param $file
+     * @return mixed
+     */
+    protected function getNewFieldValue($record, $field, $match, $file)
+    {
+        if ($file instanceof File) {
+            $newPath = $file->getURL();
+
+            $find = [
```
Comment (muskie9) [inline src/ContentFileMigrationTask.php]: Related to some notes in the comments for the PR, I've come across the `""""` issue with empty `alt` tags as well. After some testing I think removing empty `title=""` references and generating a value for empty `alt` tags might be a good option here.

**Comment by zzdjk6** [comment]
This PR looks awesome! I really need such a tool to upgrade the broken image links in `HTMLText` and `HTMLVarchar` fields.

However, if someone is in a similar situation as I am: 
- the content contains images from other places instead of being `assets`-only (such as those usage linked to `themes/`) 
- wants to keep `` tags

There is a solution here: https://github.com/zzdjk6/silverstripe-imagelinkupgrade

**Comment by muskie9** [comment]
@zzdjk6 the module looks nice. I had made some adjustments locally with this task to accommodate theme images but don't think I've had a chance to get that into this PR. I think a simple config option for retaining the `` tag rather than the new `[image]` short code would be simple enough to integrate.

**Comment by zzdjk6** [comment]
@muskie9 Well, in terms of the module task I wrote, it will modify the data in `Stage` mode, and then publish it to `Live` mode if the `Stage` version number is equal to the `Live` one (which means the staged record is already published). So basically it creates new versions and leaves the old ones unchanged.

File: src/ContentFileMigrationTask.php [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+     * @param $table
+     * @param $record
+     * @param $field
+     */
+    protected function updateFileReference($class, $table, $record, $field)
+    {
+        if (isset($record[$field])) {
+            if (preg_match('/<img\s*(?:src\s*\=\s*[\'\"](.*?)[\'\"].*?\s*|\s*|\s*)+.*?>/sm', $record[$field], $matches)) {
+                foreach ($matches as $match) {
+                    preg_match('/src=".*"/', $match, $source);
```
[the remark this answers is not in the document]
Comment (michalkleiner) [reply src/ContentFileMigrationTask.php]: Doesn't have to be named, just needs to be non-greedy.

[the remark this answers is not in the document]
Comment (wernerkrauss) [reply src/ContentFileMigrationTask.php]: Yes, it doesn't need to be named, but IMHO that's more verbose / better readable later on when we grab the $path out of the $source array. Either way, the main thing it's working like expected...

**Comment by muskie9** [comment]
Apologies for the delay in feedback implementation, circling back around on a couple projects that need this so I should be able to get this worked through over the next week.

**Comment by muskie9** [comment]
Hi @chillu, sorry for the delay, the module is probably a good idea at this point as we haven't been able to get back to this unfortunately.

File: src/ContentFileMigrationTask.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (UndefinedOffset) [reply src/ContentFileMigrationTask.php]: Ya I found `/src="([^"]*)"/` works in both places the source url is looked for in an expression, the `/src=".*"/` leaches out into other parameter :)

**Comment by chillu** [comment]
OK, as discussed with @muskie9, let's keep that code in a module and get it used by people there. @muskie9 Are you happy to host this on your own Github repo? Happy to accept a pull request linked to the module on https://docs.silverstripe.org/en/4/developer_guides/files/file_migration/ though heh. I still think this could be in core, but only with comprehensive unit tests.

**Comment by muskie9** [comment]
That all sounds good. We're happy to host the module, shooting for it being up this week with some basic usage instructions. We'll get a PR into the file migration docs page as well to go along with this. Thanks for the patience all!

**Comment by chillu** [comment]
We're working on a bugfix to redirect published files from `assets/[hash]/file.pdf` to `assets/file.pdf`, and to (optionally) provide a task to move those files to the new location as well. This was done in order to avoid broken links for sites upgrading from 3.x. This might have an overlap with the operation of this task here. https://github.com/silverstripe/silverstripe-versioned/issues/177

**Comment by purplespider** [comment]
So is this feature to rewrite image URLs to shortcodes now available as a module somewhere? Slightly confused as to its status.

**Comment by bergice** [comment]
@purplespider, please see this issue: https://github.com/silverstripe/silverstripe-assets/issues/167

It is still in progress.

**Pull request closed without merging**